src/Client.sol

```diff
@@ -6,36 +6,35 @@ import {DataCapTypes} from "filecoin-solidity/contracts/v0.8/types/DataCapTypes.
 import {DataCapAPI} from "filecoin-solidity/contracts/v0.8/DataCapAPI.sol";
 import {Errors} from "./libs/Errors.sol";
 import {CBORDecoder} from "filecoin-solidity/contracts/v0.8/utils/CborDecode.sol";
-import {IAllocation} from "./interfaces/IAllocation.sol";
+import {IClient} from "./interfaces/IClient.sol";
 
-contract Allocation is Initializable, IAllocation {
+contract Client is Initializable, IClient {
     address public manager;
-    address public client;
 
-    mapping(uint64 storageProvider => bool allowedSP) public allowedSPs;
+    mapping(address client => uint256 allowance) public clients;
+    mapping(address client => mapping(uint64 storageProvider => bool allowedSP) allowedSPs) public clientSPs;
+
 
     /// @custom:oz-upgrades-unsafe-allow constructor
     constructor() {
         _disableInitializers();
     }
 
-    function initialize(address manager_, address client_, uint64[] calldata allowedSPs_) public initializer {
+    function initialize(address manager_) public initializer {
         manager = manager_;
-        client = client_;
-        for (uint256 i = 0; i < allowedSPs_.length; i++) {
-            allowedSPs[allowedSPs_[i]] = true;
-        }
     }
 
     /**
      * @notice This function transfers DataCap tokens from the client to the storage provider
      * @dev This function can only be called by the client
      * @param params The parameters for the transfer
      */
-    function transfer(DataCapTypes.TransferParams calldata params) external onlyClient {
+    function transfer(DataCapTypes.TransferParams calldata params) external {
+        if (clients[msg.sender] == 0) revert Errors.InsufficientAllowance();
         uint64[] memory providers = _deserializeAllocationRequests(params.operator_data, 0);
         _matchAllowedSPs(providers);
         /// @custom:oz-upgrades-unsafe-allow-reachable delegatecall
+        clients[msg.sender] -= params.amount;
```

Comment:
You broke the `@custom:oz-upgrades-unsafe-allow-reachable delegatecall` annotation for transfer here.